Sources/Signal.swift

```diff
@@ -237,9 +335,67 @@ public final class Signal<Value, Error: Swift.Error> {
 	}
 }
 
-private struct SignalState<Value, Error: Swift.Error> {
-	var observers: Bag<Signal<Value, Error>.Observer> = Bag()
-	var retainedSignal: Signal<Value, Error>?
+/// The state of a `Signal`.
+///
+/// `SignalState` is guaranteed to be laid out as a tagged pointer by the Swift
+/// compiler in the support targets of the Swift 3.0.1 ABI.
+///
+/// The Swift compiler has also an optimization for enums with payloads that are
+/// all reference counted, and at most one no-payload case.
+private enum SignalState<Value, Error: Swift.Error> {
+	// FIXME: Change `()?` to `()` when concrete same-type requirement lands.
+	/// The `Signal` is alive.
+	case alive(SignalStateSnapshot<Value, Error, ()?>)
+
+	/// The `Signal` has received a termination event, and is about to be
+	/// terminated.
+	case terminating(SignalStateSnapshot<Value, Error, Event<Value, Error>>)
+
+	/// The `Signal` has terminated.
+	case terminated
+}
+
+/// The state snapshot of a `Signal` which contains the bag of observers,
+/// an optional self-retaining reference and an optional associated value.
+///
+/// As the amount of state would definitely span over a cache line,
+/// `SignalStateSnapshot` is set to be a reference type so that we can
+/// atomically update the reference instead.
+///
+/// - warning: In-place mutation should not be introduced to
+///            `SignalStateSnapshot`. Copy the states and create a new instance.
+private final class SignalStateSnapshot<Value, Error: Swift.Error, U> {
+	/// The observers of the `Signal`.
+	fileprivate let observers: Bag<Signal<Value, Error>.Observer>
+
+	/// A self-retaining reference. It is set when there are one or more active
+	/// observers.
+	fileprivate let retaining: Signal<Value, Error>?
+
+	fileprivate let associated: U
```

Comment:
Could we split `SignalStateSnapshot` into 2 different classes?

```swift
private enum SignalState<Value, Error: Swift.Error> {
    case alive(AliveState)
    case terminating(TerminatingState)
    case terminated
}

private final class AliveState {
    let observers: Bag<Signal<Value, Error>.Observer>
    let retaining: Signal<Value, Error>?
}

private final class TerminatingState {
    let observers: Bag<Signal<Value, Error>.Observer>
    let event: Event<Value, Error>
}
```

AFAICS the terminating state doesn't need `retaining` and the alive state definitely doesn't need `event`. This would simplify things a bit, I think.